service-base/src/main/java/org/eclipse/hono/service/command/CommandConnectionImpl.java

```diff
@@ -0,0 +1,126 @@
+
+import io.vertx.core.buffer.Buffer;
+import org.eclipse.hono.client.impl.HonoClientImpl;
+import org.eclipse.hono.config.ClientConfigProperties;
+import org.eclipse.hono.connection.ConnectionFactory;
+import org.eclipse.hono.util.CommandConstants;
+import org.eclipse.hono.util.ResourceIdentifier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.proton.ProtonDelivery;
+import io.vertx.proton.ProtonReceiver;
+import io.vertx.proton.ProtonSender;
+
+/**
+ * Implements a connection between an Adapter and the AMQP 1.0 network to receive commands and send a response.
+ */
+public class CommandConnectionImpl extends HonoClientImpl implements CommandConnection {
+
+    protected final Logger LOG = LoggerFactory.getLogger(getClass());
+
+    /**
+     * Creates a new client for a set of configuration properties.
+     * <p>
+     * This constructor creates a connection factory using
+     * {@link ConnectionFactory#newConnectionFactory(Vertx, ClientConfigProperties)}.
+     *
+     * @param vertx The Vert.x instance to execute the client on, if {@code null} a new Vert.x instance is used.
+     * @param clientConfigProperties The configuration properties to use.
+     * @throws NullPointerException if clientConfigProperties is {@code null}
+     */
+    public CommandConnectionImpl(final Vertx vertx, final ClientConfigProperties clientConfigProperties) {
+        super(vertx, clientConfigProperties);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public Future<Void> createCommandResponder(final String tenantId, final String deviceId,
+                                                           final Handler<Command> commandHandler) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(deviceId);
+        Objects.requireNonNull(commandHandler);
+
+        LOG.debug("create a command receiver for [tenant: {}, device-id: {}]", tenantId, deviceId);
+        Future<Void> result = Future.future();
+        connect().setHandler(h -> {
+            if(h.succeeded()) {
+                getConnection().createReceiver(ResourceIdentifier.from(CommandConstants.COMMAND_ENDPOINT, tenantId, deviceId).toString()) // TODO
+                        .openHandler(oh -> {
+                            if (oh.succeeded()) {
+                                LOG.debug("command receiver successfully opened for [tenant: {}, device-id: {}]", tenantId, deviceId);
+                                ProtonReceiver protonReceiver = oh.result();
+                                CommandResponder responder = new CommandResponder(protonReceiver);
+                                protonReceiver.handler((delivery, message) -> {
+                                    LOG.debug("command message received on [address: {}]", message.getAddress());
+                                    commandHandler.handle(new Command(responder, message));
```

Comment:
Do I see it correctly that
- the `Command` gets the responder (effectively a pair of links)
- and internally only uses the `sender` link of the responder?

I find it hard to understand here how the `Command`, the `CommandResponder` and the two Proton links interact with each other. Maybe the `Command` should only receive the sender link, since it does not use the receiver link.
In the end I do not fully understand for what the `CommandResponder` is for - maybe there is no real need for this class? WDYT?

Reply:
It is the pair of links and it implements the closing of the sender as soon as the receiver is closed. Also I could imagine to close the sender link more often - we need to test this. For me it feels natural to pair them together in a class, since they are always needed as a pair.

Reply:
My point is that I feel the current classes should be separated much better.
Looking at the code 

```
ProtonSender sender = command.getResponder().getSender();
            if (sender == null || !sender.isOpen()) {
                connect().setHandler(h -> {
                    getConnection().createSender(command.getReplyAddress())
                            .openHandler(oh -> {
                                if (oh.succeeded()) {
                                    LOG.debug("command reply sender opened successful for [replyAddress: {}]", command.getReplyAddress());
                                    command.getResponder().setSender(oh.result());
                                    command.sendResponse(data, properties, update);
                                    result.complete();
                                } else {
                                    LOG.debug("command reply failed opening for [replyAddress: {}] : {}", command.getReplyAddress(), oh.cause());
                                }
                            }).open();
                });
```

it mangles the ProtonReceiver with the Command class and also involves the CommandResponder which is changed outside of the Command class by invoking the `setSender` method and thus feels like a code smell to me.

I would at least avoid retrieving the CommandResponder and change this implementation detail of `Command` from the outside, and rather do something like
`command.setSender(oh.result());` (which internally may delegate that to the contained responder).

And if additionally the `CommandResponder` would be called something like `CommandPairManager` I would be much happier :-)